#Source
#https://benjjneb.github.io/dada2/tutorial.html
#https://benjjneb.github.io/dada2/ITS_workflow.html


# DADA2 Pipeline for Analyzing Microhaplotype NGS Data Generated by MiSeq

library(dada2)
library(ShortRead)

rm(list=ls())
path <- "D:/DADA2/MH-MiSeq" # CHANGE ME to the directory containing the fastq files after unzipping
list.files(path)

# Get forward fastq filenames
fnFs <- sort(list.files(path, pattern="_L001_R1_001.fastq", full.names = TRUE))


# Extract sample names
sample.names <- sapply(strsplit(basename(fnFs), "_L00"), `[`, 1)
sample.names

# Place filtered files in filtered subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
names(filtFs) <- sample.names


out <- filterAndTrim(fnFs, filtFs,  
                     maxN = 0, maxEE = 2, truncQ = 2, minLen = 120, 
                     rm.phix = TRUE, compress = TRUE, 
                     multithread = FALSE) # on windows, set multithread = FALSE
head(out)


errF <- learnErrors(filtFs, multithread=TRUE)


#derepFs <- derepFastq(filtFs, verbose=TRUE)
derepFs <- lapply(filtFs, derepFastq, verbose=TRUE)

#dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
dadaFs <- lapply(derepFs, dada, err=errF, multithread=TRUE)
#dadaFs[[1]]



# Convert DADA2 result to STRait Razor format
# 2023. 9. 13 ~ 15, 19, 22, 2024. 4. 24 ~ 25. 2025. 9. 11 ~ 13.


library(stringr)

configInfo <- read.table("D:/DADA2/MH24_241224.config", sep = "\t")
outputDir <- file.path(path, "Output")
if (!dir.exists(outputDir)) dir.create(outputDir)

# --- 함수 정의 ---
process_MH <- function(sample_idx, sample_name, dadaFs, configInfo, outputDir) {
  noAllele <- length(dadaFs[[sample_idx]]$denoised)
  countMH  <- nrow(configInfo)
  
  filePath <- file.path(outputDir, paste0(sample_name, "_DADA2.txt"))
  file.create(filePath)  # 파일 초기화
  
  for (j in seq_len(countMH)) {
    MH_marker  <- configInfo[[1]][j]
    fwd_primer <- configInfo[[3]][j]
    rev_primer <- configInfo[[4]][j]
    
    # --- allele 정보를 모아두는 데이터프레임 ---
    alleleInfo <- data.frame(
      markerNm  = character(0),
      length_bp = integer(0),
      sequence  = character(0),
      abundance = integer(0),
      stringsAsFactors = FALSE
    )
    
    for (k in seq_len(noAllele)) {
      alleleSeq <- dadaFs[[sample_idx]]$sequence[k]
      abundance <- dadaFs[[sample_idx]]$denoised[k]
      
      fwdPos <- str_locate(alleleSeq, fwd_primer)
      revPos <- str_locate(alleleSeq, rev_primer)
      if (all(!is.na(fwdPos), !is.na(revPos))) {
        targetSeq <- str_sub(alleleSeq, start = fwdPos[2] + 1, end = revPos[1] - 1)
        targetLen <- revPos[1] - fwdPos[2] - 1
        
        alleleInfo <- rbind(
          alleleInfo,
          data.frame(
            markerNm  = MH_marker,
            length_bp = targetLen,
            sequence  = targetSeq,
            abundance = abundance,
            stringsAsFactors = FALSE
          )
        )
      } else if (any(!is.na(fwdPos), !is.na(revPos)) && abundance > 100) {
    		message(sprintf("Chimera? Al:%d | %s | %d %s...",
                        k, MH_marker, abundance, substr(alleleSeq, 1, 50)))
      }
    }
    
    # --- 동일한 sequence 합치기 ---
    if (nrow(alleleInfo) > 0) {
      alleleInfo <- aggregate(abundance ~ markerNm + length_bp + sequence,
                              data = alleleInfo, FUN = sum)

      # → abundance 기준 내림차순 정렬
      alleleInfo <- alleleInfo[order(-alleleInfo$abundance), ]
            
      # alleleLines로 변환
      alleleLines <- sprintf(
        "%s\t%d bases\t%s\t%d\t0",
        alleleInfo$markerNm,
        alleleInfo$length_bp,
        alleleInfo$sequence,
        alleleInfo$abundance
      )
    } else {
      alleleLines <- sprintf(
        "%s\t0 bases\tSumBelowThreshold\t0\t0",
        MH_marker)
    }
    
    # --- 마커별 기록 후 빈 줄 추가 ---
    cat(alleleLines, file=filePath, sep="\n", append=TRUE)
    cat("\n", file=filePath, append=TRUE)
  }
  
  message("→ saved to ", filePath)
}

# --- 실행 부분 ---
for (i in seq_along(sample.names)) {
  cat(sample.names[i], "is converting...\n")
  process_MH(i, sample.names[i], dadaFs, configInfo, outputDir)
}
