# DADA2-SE

![Last Updated](https://img.shields.io/github/last-commit/Microhap/DADA2-SE?label=last%20updated)

DADA2-SE is an R script that performs denoising of single-end Illumina MiSeq sequencing data for microhaplotypes using DADA2. 
<br> <br> 
The denoising pipeline for microhaplotype consists of three main steps: ASV inference, haplotype identification, and genotype calling. This script performs the ASV inference and haplotype identification steps. The subsequent genotype calling step is performed by [Visual Microhap](http://forensic.yonsei.ac.kr/VisualMH/index.html). 

<br> 

![Denoising Pipeline](image/denoising%20pipeline.PNG)
 
<br>

## Requirements
- R version >= 4.0
- Required packages: `dada2`, `ShortRead`, `stringr`

<br>

If these packages are not installed, please install them before running the script. 

```
# Bioconductor
install.packages("BiocManager")
BiocManager::install(c("dada2", "ShortRead"))

# CRAN
install.packages("stringr")
```
<br>

## Inputs

### 1.  FASTQ files 
This pipeline recommends using the forward reads as input. Merged paired-end reads into a single FASTQ file are not recommended due to potential quality issues in reverse reads. The FASTQ files should be generated from the same sequencing batch to ensure consistent error modeling and quality control.

<br>

Make sure to modify the path in the script to reflect the absolute path of your FASTQ files.

```
path <- "D:/DADA2-SE-main/test/"   # CHANGE ME to the directory containing the fastq files after unzipping
```

<br>

### 2.  Configuration file 
Haplotype identification was performed by referencing a configuration file containing marker-specific sequences. The configuration file uses the same format as that of STRait Razor 3.0.

<br>

Below is an example of a configuration file:

| Marker      | Type | 5'Flank        | 3'Flank        | Motif | Period | Offset |
|-------------|------|----------------|----------------|-------|--------|--------|
| mh09KK_153  | MH   | GGCAGTCTTCAT   | GGCTGTATGATA   | A     | 1      | 155    |
| mh02KK_134  | MH   | CCCTTGGCAGGA   | TACGAGCCTCAA   | A     | 1      | 118    |
| mh13KK_213  | MH   | CAGCAAGGAGAA   | TTGAGTTGATCA   | A     | 1      | 194    |

 - **Marker**: The identifier for the microhaplotype marker (e.g., mh09KK-153).
 - **Type**: The type of marker ("MH" for microhaplotypes).
 - **5'Flank**: Part of the 5'-end sequence of each marker.
 - **3'Flank**: Part of the 3'-end sequence of each marker.
 - **Motif**: The repeat motif of the marker, set to "A" in this microhaplotype analysis.
 - **Period**: The number of bases in the repeat unit of the motif, set to "1" in this microhaplotype analysis.
 - **Offset**: The length of the sequence inside the 5'Flank and 3'Flank regions.

<br> 

Make sure to modify the path in the script to reflect the absolute path of your configuration file.

```
configInfo <- read.table("D:/DADA2-SE-main/test/MH24_241224.config", sep = "\t")   # CHANGE ME to the directory containing Configuration file 
```

<br>

## Test dataset 
You can download the test dataset in the `test/` directory. 

<br>

The `test/` directory includes:

- Two forward FASTQ files containing amplicon sequencing data for 24 microhaplotypes from the 2800M reference DNA:
```
./test/2800M#1_S1_L001_R1_001.fastq
```
```
./test/2800M#2_S2_L001_R1_001.fastq
```

<br>

- A configuration file for 24 microhaplotypes:
```
./test/MH24_241224.config
```


<br>

## Outputs
All output files are automatically generated in an **Output** subdirectory under the user-defined path.
The Output directory is created automatically if it does not already exist.

<br>

Final output files will be saved to:
```
./test/Output/
```

<br>

Output contains denoised haplotype data for each marker.

<br>

Below is an example of the output for 2800M#1 generated by the project:

| Marker      | Size(bp) | Sequence                                                                                                                                                                                                                                      | Coverage#1 | Coverage#2 |
|-------------|----------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------|------|
| mh09KK_153  | 155            | GCTCTGAACAATTGGGTATTCTTTTTTCTTAGAGCCCAGATGCATTTTTTTGAAAGTCGTTCCAGGGGCCTGAGATGAAGTGGGGGTGTGAGAAGTAAGTTGGCTAGGGCAGATAGAACCTAAGTGTCTTCTCCTTAAGTCAGCTCCCCTTATGA                                     | 1677          | 0    |
| mh09KK_153  | 155            | GCTCTGAACAATTGGGTATTCTTTTTTCTTAGAGCCCAGATGCATTTTTTTGAAAGTCGTTCCAGGGGCCTGAGATGAAGTGGGGGTGTGAGAAGTAAGTTGGCTAGGGCAGATAGCACCTAAGTGTCTTCTCCTTAAGTCAGCTCCCCTTATGA                                     | 1186          | 0    |
| mh02KK_134  | 118            | ACCCTCACTACCTAAGGATGGGCAATGGCTCATGAGTGAGAAACATGGAGCCGTGGGAACTCAGAATGACATGCTACCTGGAGATTGTGGTAACGCCCTGTTTTTTTGTGGGCATATC                                                                                   | 1651          | 0    |
| mh02KK_134  | 118            | ACCCTCTCTACCTAAGGATGGGCAATGGCTTATGAGTGAGAAACATGGAGCCGTGGGAACTCAGAATGACATGCTACCTGGAGATTGTGGTAACGCCCTGTTTTTTTGTGGGCATATC                                                                                   | 1429          | 0    |
| mh13KK_213  | 195            | CTTCAGTTGTCAAGGTATTGGGTACAGGGGTCAGAAAGAAACATGACTCCATGGACCACTGCTTGGCCCAAGACCAGATGTCAAAACCACAGAGCCCTGCTGTAGAGCATTACAAATGTATTCCACCAAATGTTGGGATGCATCCTAGACCTGTGCTGACCAGCAGTCCCCAGCTGTGAGGAGAAGCCCGCCATT | 2133          | 0    |


It can be directly used as input for [Visual Microhap](http://forensic.yonsei.ac.kr/VisualMH/index.html), enabling genotype calling of microhaplotype.


<br>

## Citation
If you use this pipeline in your work, please cite the following article:

> *Y.L. Kwon, J. Kim, S.M. Joo, K.J. Shin*  
> **_Denoising of microhaplotype MPS data using DADA2 and its application to two-person DNA mixture analysis_**  
> *Forensic Science International: Genetics, 2025, 103295.*  
> [https://doi.org/10.1016/j.fsigen.2025.103295](https://doi.org/10.1016/j.fsigen.2025.103295)



